R/FilterStateChoices.R

```diff
@@ -136,6 +136,98 @@ ChoicesFilterState <- R6::R6Class( # nolint
     #' filter$set_selected(c("c", "a"))
     set_selected = function(value) {
       super$set_selected(value)
+    },
+
+    #' @description
+    #' Shiny module server.
+    #'
+    #' @param id (`character(1)`)\cr
+    #'   shiny module instance id
+    #'
+    #' @return `moduleServer` function which returns reactive value
+    #'   signaling that remove button has been clicked
+    #'
+    server = function(id) {
+      moduleServer(
+        id = id,
+        function(input, output, session) {
+
+          observeEvent(self$get_keep_na(), {
+            
+            if (self$get_keep_na()) {
+                class <- "fa fa-check"
+              } else {
+                class <- "fa fa-xmark"
+              }
+
+            output$header_keep_na <- renderUI({
+              tagList(
+                tags$strong("NA "),
+                tags$span(class = class)
+              )
+            })
+          })
+
+          observeEvent(self$get_selected(), {
+            if (length(self$get_selected() > 1)) {
+              value <- paste0(
+                "(",
+                paste(self$get_selected(), collapse = "-"),
+                ")"
+              )
+            } else {
+              value <- self$get_selected()
+            }
+
+            output$header_name_value <- renderUI({
+              tagList(
+                tags$strong(self$get_varname(deparse = TRUE)),
+                tags$span(value)
+              )
+            })
+          })
+          private$server_inputs("inputs")
+          reactive(input$remove) # back to parent to remove self
+        }
+      )
+    },
+
+    #' @description
+    #' Shiny module UI.
+    #'
+    #' @param id (`character(1)`)\cr
+    #'  shiny element (module instance) id;
+    #'  the UI for this class contains simple message stating that it is not supported
+    #'
+    ui = function(id) {
+      ns <- NS(id)
+      
+      tags$li(
+        id = id,
+        tags$div(
+        class = "filter-card filter-card-range",
+        tags$div(
+          class = "filter-card-header",
+          uiOutput(ns("header_name_value"), inline = TRUE),
+          uiOutput(ns("header_keep_na"), inline = TRUE),
+          tags$div(
+            class = "filter-card-icons",
+            tags$span(
+              class = "filter-card-toggle fa fa-chevron-right"
+            ),
+            actionLink(
+              inputId = ns("remove"),
+              label = icon("circle-xmark", lib = "font-awesome"),
+              class = "filter-card-remove"
+            )
+          )
+        ),
+        tags$div(
+          class = "filter-card-body",
+          private$ui_inputs(ns("inputs"))
+        )
+        )
+      )
```

Comment:
I think we can keep ui in the FilterState as before but header could be delegated to the child-classes - in the FIlterState it could look like this:
```r
ui = div(
  class = "filter-card-header",
  private$ui_header(ns("header")),
  actionLink(ns("remove"),...)
)
div(
  private$ui_inputs(ns("inputs"))
)
```

Reply:
I also see reusable code in server method. Maybe we can move that to FilterState class too?

Reply:
Definitely some repeated code. I think I misunderstood a conversation the other day.